File: app/src/main/kotlin/com/przunk/protracktor/data/SongLengths.kt
/*
 * Protracktor -- a player for retro platform music formats.
 * Copyright (C) 2026 Przunk
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See
 * the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If
 * not, see <https://www.gnu.org/licenses/>.
 */
package com.przunk.protracktor.data

/**
 * HVSC's song length database, as parsing and nothing else.
 *
 * A SID file carries no duration. The tune is a program: it plays until someone stops it, and
 * "how long is it" is a question the format cannot answer. The High Voltage SID Collection answers
 * it by hand -- a person listened and wrote a number down -- and publishes the result as
 * `DOCUMENTS/Songlengths.md5`. Without it every SID in the app shows an unknown length and the
 * position readout has nothing to count towards.
 *
 * **No Android imports on purpose**, the same reason as [SchemaSql]: 5 MB of somebody else's text
 * format is exactly the thing to test on the JVM rather than discover on a phone.
 *
 * ### The file
 *
 * ```
 * [Database]
 * ; /MUSICIANS/H/Hubbard_Rob/Commando.sid
 * 6d019ecba831a9f853675aac29a61c10=3:55.594 1:01.288 0:06 ...
 * ```
 *
 * The key is the **plain MD5 of the whole SID file**. That is worth stating because it used not to
 * be: older HVSC releases keyed the database on a hash built from selected header fields, which
 * libsidplayfp still exposes as `SidTune::createMD5`. Checked against the real file on 2026-09-02 --
 * three tunes fetched from the collection, all three found by plain file MD5 -- so nothing here
 * needs the SID parser at all, and a lookup costs one hash of bytes we are holding anyway.
 *
 * One line per tune, one time per subsong, `M:SS` or `M:SS.fff` where the fraction is a decimal
 * fraction of a second. Measured over the whole 5.2 MB file on 2026-09-02: 61,157 entries, no
 * duplicate keys, no token of any other shape, at most 256 subsongs on one line, longest tune
 * 33:46.
 */
object SongLengths {

    /** One tune: the file's MD5, and a length in seconds per subsong. */
    data class Entry(val md5: String, val seconds: List<Double>)

    private val ENTRY = Regex("^([0-9a-fA-F]{32})=(.+)$")
    private val TIME = Regex("^(\\d+):([0-5]\\d)(?:\\.(\\d{1,3}))?$")

    /**
     * Reads the database.
     *
     * Anything that is not an entry line is skipped rather than rejected -- the `[Database]` header
     * and a comment naming the path above every single entry are both expected, and a future
     * release adding a section this does not know about should cost the lengths it does know, not
     * all of them. A malformed time inside an otherwise good line drops that **line**, because a
     * tune whose third subsong is unreadable is a tune whose lengths we cannot trust.
     */
    fun parse(text: String): List<Entry> = text.lineSequence().mapNotNull { line ->
        val match = ENTRY.matchEntire(line.trim()) ?: return@mapNotNull null
        val times = match.groupValues[2].trim().split(' ').filter { it.isNotEmpty() }
        val seconds = times.map { parseTime(it) ?: return@mapNotNull null }
        if (seconds.isEmpty()) null else Entry(match.groupValues[1].lowercase(), seconds)
    }.toList()

    /** `3:55.594` to 235.594. Null if it is not a time, which is the caller's cue to drop the line. */
    fun parseTime(token: String): Double? {
        val match = TIME.matchEntire(token) ?: return null
        val minutes = match.groupValues[1].toLongOrNull() ?: return null
        val fraction = match.groupValues[3]
        // "3.5" is three and a half seconds. Spelling it out because the field is described as
        // milliseconds in places, and reading .5 as five milliseconds would make every fractional
        // length wrong by up to a second in a way nobody would notice from the numbers.
        val frac = if (fraction.isEmpty()) 0.0 else fraction.toInt() / TENS[fraction.length]
        return minutes * 60.0 + match.groupValues[2].toInt() + frac
    }

    private val TENS = doubleArrayOf(1.0, 10.0, 100.0, 1000.0)

    /** How the seconds go into one column, and come back out. Space-separated, as the file has them. */
    fun pack(seconds: List<Double>): String = seconds.joinToString(" ") { trimZeros(it) }

    fun unpack(packed: String): List<Double> =
        packed.split(' ').mapNotNull { it.toDoubleOrNull() }

    private fun trimZeros(value: Double): String {
        val whole = value.toLong()
        return if (value == whole.toDouble()) whole.toString() else value.toString()
    }
}

File: web/src/songlengths.js
// SPDX-FileCopyrightText: 2026 Przunk
// SPDX-License-Identifier: GPL-3.0-or-later

// HVSC's song lengths, for the page.
//
// **The phone's `data/SongLengths.kt`, in the other language** (`docs/STATUS.md` C56). A SID file
// carries no duration: the tune is a program, it plays until somebody stops it, and "how long is
// it" is a question the format cannot answer. The High Voltage SID Collection answers it by hand --
// a person listened and wrote a number down -- and publishes the result as
// `DOCUMENTS/Songlengths.md5`, keyed by the plain MD5 of the whole file.
//
// Without it the page played every SID for ever, and the phone did too until its database was
// downloaded. The fallback length added for C56 stays either way: it is the answer for every format
// nobody has measured, and HVSC only knows about the C64.
//
// The time tokens are checked against `docs/rules/queue-cases.tsv` on both sides, so the two cannot
// drift. The fraction is the trap that file exists for.

/**
 * `3:55.594` to 235.594, or null when the token is not a time.
 *
 * Null is the caller's cue to drop the **whole line**: a tune whose third subsong is unreadable is
 * a tune whose lengths cannot be trusted.
 */
export function parseSongLengthTime(token) {
  const match = /^(\d+):([0-5]\d)(?:\.(\d{1,3}))?$/.exec(token);
  if (!match) return null;
  const [, minutes, seconds, fraction] = match;
  // **A decimal fraction of a second, not milliseconds.** ".5" is half a second. The field is
  // described as milliseconds in places, and reading it that way makes every fractional length
  // wrong by up to a second -- in a way nobody would notice from the numbers.
  const frac = fraction ? Number(fraction) / 10 ** fraction.length : 0;
  return Number(minutes) * 60 + Number(seconds) + frac;
}

/**
 * Reads the database into `{ md5, seconds }` entries.
 *
 * Anything that is not an entry line is skipped rather than rejected. The `[Database]` header and a
 * comment naming the path above every entry are both expected, and a future release adding a
 * section this does not know about should cost the lengths it does know rather than all of them.
 */
export function parseSongLengths(text) {
  const entries = [];
  for (const raw of text.split('\n')) {
    const line = raw.trim();
    const match = /^([0-9a-fA-F]{32})=(.+)$/.exec(line);
    if (!match) continue;
    const tokens = match[2].trim().split(' ').filter(Boolean);
    const seconds = [];
    let good = tokens.length > 0;
    for (const token of tokens) {
      const value = parseSongLengthTime(token);
      if (value == null) { good = false; break; }
      seconds.push(value);
    }
    if (good) entries.push({ md5: match[1].toLowerCase(), seconds });
  }
  return entries;
}

// --- MD5 ----------------------------------------------------------------------------------------
//
// **Written out rather than fetched from anywhere**, and it is the one thing on this page that had
// to be: `crypto.subtle` does not offer MD5 and never will, because MD5 is broken as a *security*
// hash. Nothing here is security -- HVSC chose it as a **key**, twenty years ago, and the lookup
// has to use the key the database was written with or it finds nothing.
//
// Sixty lines, no dependency, and checked against the published vectors in `check-page.mjs`.

const S = [
  7, 12, 17, 22, 7, 12, 17, 22, 7, 12, 17, 22, 7, 12, 17, 22,
  5, 9, 14, 20, 5, 9, 14, 20, 5, 9, 14, 20, 5, 9, 14, 20,
  4, 11, 16, 23, 4, 11, 16, 23, 4, 11, 16, 23, 4, 11, 16, 23,
  6, 10, 15, 21, 6, 10, 15, 21, 6, 10, 15, 21, 6, 10, 15, 21,
];

/** floor(abs(sin(i + 1)) * 2^32), the constants the algorithm is defined with. */
const K = new Uint32Array(64);
for (let i = 0; i < 64; i++) K[i] = Math.floor(Math.abs(Math.sin(i + 1)) * 4294967296);

const rotate = (value, by) => (value << by) | (value >>> (32 - by));

/**
 * The MD5 of some bytes, lower-case hex.
 *
 * Takes a `Uint8Array` -- the page holds a tune as one, having just fetched it -- and pads into a
 * copy rather than growing the caller's buffer.
 */
export function md5(bytes) {
  const length = bytes.length;
  // The message, a 0x80 byte, zeros to 56 mod 64, then the bit length as 64 little-endian bits.
  const padded = new Uint8Array(((length + 8) >> 6 << 6) + 64);
  padded.set(bytes);
  padded[length] = 0x80;
  const bits = length * 8;
  const view = new DataView(padded.buffer);
  view.setUint32(padded.length - 8, bits >>> 0, true);
  // **The high word matters**, because a 5 MB file is 41 million bits and `bits >>> 0` keeps only
  // 32 of them. Written as a float division so it stays exact past 2^32.
  view.setUint32(padded.length - 4, Math.floor(bits / 4294967296), true);

  let [a0, b0, c0, d0] = [0x67452301, 0xefcdab89, 0x98badcfe, 0x10325476];
  const chunk = new Uint32Array(16);

  for (let at = 0; at < padded.length; at += 64) {
    for (let i = 0; i < 16; i++) chunk[i] = view.getUint32(at + i * 4, true);
    let [a, b, c, d] = [a0, b0, c0, d0];
    for (let i = 0; i < 64; i++) {
      let f;
      let g;
      if (i < 16) { f = (b & c) | (~b & d); g = i; }
      else if (i < 32) { f = (d & b) | (~d & c); g = (5 * i + 1) % 16; }
      else if (i < 48) { f = b ^ c ^ d; g = (3 * i + 5) % 16; }
      else { f = c ^ (b | ~d); g = (7 * i) % 16; }
      f = (f + a + K[i] + chunk[g]) | 0;
      a = d;
      d = c;
      c = b;
      b = (b + rotate(f, S[i])) | 0;
    }
    a0 = (a0 + a) | 0;
    b0 = (b0 + b) | 0;
    c0 = (c0 + c) | 0;
    d0 = (d0 + d) | 0;
  }

  const out = new Uint8Array(16);
  new DataView(out.buffer).setUint32(0, a0 >>> 0, true);
  new DataView(out.buffer).setUint32(4, b0 >>> 0, true);
  new DataView(out.buffer).setUint32(8, c0 >>> 0, true);
  new DataView(out.buffer).setUint32(12, d0 >>> 0, true);
  return [...out].map((byte) => byte.toString(16).padStart(2, '0')).join('');
}
